jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/internal/FrameFlusher.java

```diff
@@ -352,104 +361,77 @@ protected Action process() throws Throwable
+            return _queue.size();
         }
     }
 
-    private void timeoutExpired()
+    @Override
+    protected void onSuccess()
     {
-        boolean failed = false;
-        try (AutoLock l = lock.lock())
+        List<Entry> succeededEntries;
+        try (AutoLock l = _lock.lock())
         {
-            if (closedCause != null)
-                return;
-
-            long currentTime = System.currentTimeMillis();
-            long expiredIfCreatedBefore = currentTime - idleTimeout;
-            long earliestEntry = currentTime;
-
-            /* Iterate through entries in both the queue and entries list.
-            If any entry has expired then we fail the FrameFlusher.
-            Otherwise we will try to schedule a new timeout. */
-            Iterator<Entry> iterator = TypeUtil.concat(entries.iterator(), queue.iterator());
-            while (iterator.hasNext())
-            {
-                Entry entry = iterator.next();
-
-                if (entry.getTimeOfCreation() <= expiredIfCreatedBefore)
-                {
-                    LOG.warn("FrameFlusher write timeout on entry: {}", entry);
-                    failed = true;
-                    canEnqueue = false;
-                    closedCause = new WebSocketWriteTimeoutException("FrameFlusher Write Timeout");
-                    failedEntries.addAll(entries);
-                    failedEntries.addAll(queue);
-                    entries.clear();
-                    queue.clear();
-                    break;
-                }
-
-                if (entry.getTimeOfCreation() < earliestEntry)
-                    earliestEntry = entry.getTimeOfCreation();
-            }
+            succeededEntries = new ArrayList<>(_entries);
+            _entries.clear();
+        }
 
-            // if a timeout is set schedule a new timeout if we haven't failed and still have entries
-            if (!failed && idleTimeout > 0 && !(entries.isEmpty() && queue.isEmpty()))
-            {
-                long nextTimeout = earliestEntry + idleTimeout - currentTime;
-                cyclicTimeout.schedule(nextTimeout, TimeUnit.MILLISECONDS);
-            }
+        for (Entry entry : succeededEntries)
+        {
+            if (entry.frame.getOpCode() == OpCode.CLOSE)
+                _endPoint.shutdownOutput();
+            notifyCallbackSuccess(entry.callback);
         }
 
-        if (failed)
-            this.iterate();
+        _releasableBuffers.forEach(RetainableByteBuffer::release);
+        _releasableBuffers.clear();
+        super.onSuccess();
     }
 
     @Override
     public void onCompleteFailure(Throwable failure)
     {
-        if (batchBuffer != null)
-            batchBuffer.clear();
-        releaseAggregate();
-        try (AutoLock l = lock.lock())
+        if (_batchBuffer != null)
+            _batchBuffer.clear();
+        releaseAggregateIfEmpty();
+        List<Entry> failedEntries;
+        try (AutoLock l = _lock.lock())
```

Comment:
Currently since iterate grabs the lock, the lock will already be held here.   That means we are calling notify below with the lock held, which I don't think we should do.